### PR TITLE
Add argstream module

### PR DESCRIPTION
**Pull request by jcorbin**
First step towards implementing streaming.

This module is the core piece that knows how to split/combine arg chunks from/to arg streams.

Test is an exhaustive integration test, the current search space covered is 649 cases, however it still gets to run fast since `setImmediate` deferral is sufficient "asynchrony" for this test.

reviewers: @Raynos @kriskowal

### CODE REVIEW

**Comment by jcorbin** [comment]
Currently the test hangs when run from `test/index.js` ala `npm test` even tho running `node test/argstream.js` directly works fine.

File: node/argstream.js [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+    self.arg3.on('error', passError);
+    function passError(err) {
+        self.emit('error', err);
+    }
+
+    self.arg2.on('start', function onArg2Start() {
+        if (!self.arg1._writableState.ended) self.arg1.end();
+    });
+    self.arg3.on('start', function onArg3Start() {
+        if (!self.arg2._writableState.ended) self.arg2.end();
```
Comment (Raynos) [inline node/argstream.js]: Is this a programmer error ?

It seems like cutting arg2 mid stream when someone writes to arg3 is going to lead to corrupted data somewhere.

File: node/argstream.js [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+        if (++self._numFinished === 3) {
+            self.finished = true;
+            self.emit('finish');
+        }
+    }
+}
+
+inherits(InArgStream, ArgStream);
+
+InArgStream.prototype.handleFrame = function handleFrame(parts) {
```
Comment (Raynos) [inline node/argstream.js]: Is `parts` a triplet of buffers, i.e. `[arg1, arg2, arg3]` ?

**Comment by kriskowal** [comment]
I like how compartmentalized this commit is. I expect that with this factoring we will be able to use the arg stream object in the Autobahn clients without imposing it upon the Autobahn service.

:ok:

**Comment by kriskowal** [comment]
:ok:

**Pull request merged**